File: java/src/main/java/com/databricks/zerobus/ArrowStreamConfigurationOptions.java
package com.databricks.zerobus;

/**
 * Configuration options for Arrow Flight streams.
 *
 * <p>This class provides settings to control Arrow Flight stream behavior including performance
 * tuning, recovery, and connection management.
 *
 * <p>Use the builder pattern to create instances:
 *
 * <pre>{@code
 * ArrowStreamConfigurationOptions options = ArrowStreamConfigurationOptions.builder()
 *     .setMaxInflightBatches(2000)
 *     .setRecovery(true)
 *     .setFlushTimeoutMs(600000)
 *     .build();
 * }</pre>
 *
 * @see ZerobusArrowStream
 * @see ZerobusSdk#createArrowStream
 */
public class ArrowStreamConfigurationOptions {

  private int maxInflightBatches = 1000;
  private boolean recovery = true;
  private long recoveryTimeoutMs = 15000;
  private long recoveryBackoffMs = 2000;
  private int recoveryRetries = 4;
  private long serverLackOfAckTimeoutMs = 60000;
  private long flushTimeoutMs = 300000;
  private long connectionTimeoutMs = 30000;

  private ArrowStreamConfigurationOptions() {}

  private ArrowStreamConfigurationOptions(
      int maxInflightBatches,
      boolean recovery,
      long recoveryTimeoutMs,
      long recoveryBackoffMs,
      int recoveryRetries,
      long serverLackOfAckTimeoutMs,
      long flushTimeoutMs,
      long connectionTimeoutMs) {
    this.maxInflightBatches = maxInflightBatches;
    this.recovery = recovery;
    this.recoveryTimeoutMs = recoveryTimeoutMs;
    this.recoveryBackoffMs = recoveryBackoffMs;
    this.recoveryRetries = recoveryRetries;
    this.serverLackOfAckTimeoutMs = serverLackOfAckTimeoutMs;
    this.flushTimeoutMs = flushTimeoutMs;
    this.connectionTimeoutMs = connectionTimeoutMs;
  }

  /**
   * Returns the maximum number of batches that can be in flight.
   *
   * <p>This controls how many batches the SDK can accept and send to the server before waiting for
   * acknowledgments. Higher values improve throughput but use more memory.
   *
   * @return the maximum number of in-flight batches
   */
  public int maxInflightBatches() {
    return this.maxInflightBatches;
  }

  /**
   * Returns whether automatic recovery is enabled.
   *
   * <p>When enabled, the SDK will automatically attempt to recover from stream failures by
   * reconnecting and resending unacknowledged batches.
   *
   * @return true if automatic recovery is enabled, false otherwise
   */
  public boolean recovery() {
    return this.recovery;
  }

  /**
   * Returns the timeout for recovery operations.
   *
   * @return the recovery timeout in milliseconds
   */
  public long recoveryTimeoutMs() {
    return this.recoveryTimeoutMs;
  }

  /**
   * Returns the backoff delay between recovery attempts.
   *
   * @return the recovery backoff delay in milliseconds
   */
  public long recoveryBackoffMs() {
    return this.recoveryBackoffMs;
  }

  /**
   * Returns the maximum number of recovery attempts.
   *
   * @return the maximum number of recovery attempts
   */
  public int recoveryRetries() {
    return this.recoveryRetries;
  }

  /**
   * Returns the timeout for server acknowledgment.
   *
   * @return the server acknowledgment timeout in milliseconds
   */
  public long serverLackOfAckTimeoutMs() {
    return this.serverLackOfAckTimeoutMs;
  }

  /**
   * Returns the timeout for flush operations.
   *
   * @return the flush timeout in milliseconds
   */
  public long flushTimeoutMs() {
    return this.flushTimeoutMs;
  }

  /**
   * Returns the timeout for establishing a connection.
   *
   * @return the connection timeout in milliseconds
   */
  public long connectionTimeoutMs() {
    return this.connectionTimeoutMs;
  }

  /**
   * Returns the default Arrow stream configuration options.
   *
   * <p>Default values:
   *
   * <ul>
   *   <li>maxInflightBatches: 1000
   *   <li>recovery: true
   *   <li>recoveryTimeoutMs: 15000
   *   <li>recoveryBackoffMs: 2000
   *   <li>recoveryRetries: 4
   *   <li>serverLackOfAckTimeoutMs: 60000
   *   <li>flushTimeoutMs: 300000
   *   <li>connectionTimeoutMs: 30000
   * </ul>
   *
   * @return the default Arrow stream configuration options
   */
  public static ArrowStreamConfigurationOptions getDefault() {
    return new ArrowStreamConfigurationOptions();
  }

  /**
   * Returns a new builder for creating ArrowStreamConfigurationOptions.
   *
   * @return a new builder
   */
  public static ArrowStreamConfigurationOptionsBuilder builder() {
    return new ArrowStreamConfigurationOptionsBuilder();
  }

  /**
   * Builder for creating {@link ArrowStreamConfigurationOptions} instances.
   *
   * <p>All parameters have sensible defaults if not specified.
   */
  public static class ArrowStreamConfigurationOptionsBuilder {
    private final ArrowStreamConfigurationOptions defaults =
        ArrowStreamConfigurationOptions.getDefault();

    private int maxInflightBatches = defaults.maxInflightBatches;
    private boolean recovery = defaults.recovery;
    private long recoveryTimeoutMs = defaults.recoveryTimeoutMs;
    private long recoveryBackoffMs = defaults.recoveryBackoffMs;
    private int recoveryRetries = defaults.recoveryRetries;
    private long serverLackOfAckTimeoutMs = defaults.serverLackOfAckTimeoutMs;
    private long flushTimeoutMs = defaults.flushTimeoutMs;
    private long connectionTimeoutMs = defaults.connectionTimeoutMs;

    private ArrowStreamConfigurationOptionsBuilder() {}

    /**
     * Sets the maximum number of batches that can be in flight.
     *
     * @param maxInflightBatches the maximum number of in-flight batches
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setMaxInflightBatches(int maxInflightBatches) {
      this.maxInflightBatches = maxInflightBatches;
      return this;
    }

    /**
     * Sets whether automatic recovery is enabled.
     *
     * @param recovery true to enable automatic recovery, false to disable
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setRecovery(boolean recovery) {
      this.recovery = recovery;
      return this;
    }

    /**
     * Sets the timeout for recovery operations.
     *
     * @param recoveryTimeoutMs the recovery timeout in milliseconds
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setRecoveryTimeoutMs(long recoveryTimeoutMs) {
      this.recoveryTimeoutMs = recoveryTimeoutMs;
      return this;
    }

    /**
     * Sets the backoff delay between recovery attempts.
     *
     * @param recoveryBackoffMs the recovery backoff delay in milliseconds
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setRecoveryBackoffMs(long recoveryBackoffMs) {
      this.recoveryBackoffMs = recoveryBackoffMs;
      return this;
    }

    /**
     * Sets the maximum number of recovery attempts.
     *
     * @param recoveryRetries the maximum number of recovery attempts
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setRecoveryRetries(int recoveryRetries) {
      this.recoveryRetries = recoveryRetries;
      return this;
    }

    /**
     * Sets the timeout for server acknowledgment.
     *
     * @param serverLackOfAckTimeoutMs the server acknowledgment timeout in milliseconds
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setServerLackOfAckTimeoutMs(
        long serverLackOfAckTimeoutMs) {
      this.serverLackOfAckTimeoutMs = serverLackOfAckTimeoutMs;
      return this;
    }

    /**
     * Sets the timeout for flush operations.
     *
     * @param flushTimeoutMs the flush timeout in milliseconds
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setFlushTimeoutMs(long flushTimeoutMs) {
      this.flushTimeoutMs = flushTimeoutMs;
      return this;
    }

    /**
     * Sets the timeout for establishing a connection.
     *
     * @param connectionTimeoutMs the connection timeout in milliseconds
     * @return this builder for method chaining
     */
    public ArrowStreamConfigurationOptionsBuilder setConnectionTimeoutMs(long connectionTimeoutMs) {
      this.connectionTimeoutMs = connectionTimeoutMs;
      return this;
    }

    /**
     * Builds a new ArrowStreamConfigurationOptions instance.
     *
     * @return a new ArrowStreamConfigurationOptions with the configured settings
     */
    public ArrowStreamConfigurationOptions build() {
      return new ArrowStreamConfigurationOptions(
          this.maxInflightBatches,
          this.recovery,
          this.recoveryTimeoutMs,
          this.recoveryBackoffMs,
          this.recoveryRetries,
          this.serverLackOfAckTimeoutMs,
          this.flushTimeoutMs,
          this.connectionTimeoutMs);
    }
  }
}

File: dotnet/src/Databricks.Zerobus/ArrowStreamConfigurationOptions.cs
namespace Databricks.Zerobus;

/// <summary>
/// Configuration options for Zerobus Arrow Flight streams (Beta).
/// Extends gRPC stream options with Arrow-specific settings.
/// </summary>
public sealed class ArrowStreamConfigurationOptions
{
    /// <summary>
    /// Default: 1,000 batches.
    /// </summary>
    public const int DefaultMaxInflightBatches = 1_000;

    /// <summary>
    /// Default: 30,000 ms (30 seconds).
    /// </summary>
    public const int DefaultConnectionTimeoutMs = 30_000;

    /// <summary>
    /// Default: -1 (wait full server-specified duration).
    /// </summary>
    public const long DefaultStreamPausedMaxWaitTimeMs = -1;

    /// <summary>
    /// Maximum number of Arrow batches that can be in flight.
    /// </summary>
    public int MaxInflightBatches { get; }

    /// <summary>
    /// Whether automatic recovery is enabled. Default: true.
    /// </summary>
    public bool Recovery { get; }

    /// <summary>
    /// Recovery timeout in milliseconds.
    /// </summary>
    public int RecoveryTimeoutMs { get; }

    /// <summary>
    /// Recovery backoff delay in milliseconds.
    /// </summary>
    public int RecoveryBackoffMs { get; }

    /// <summary>
    /// Maximum number of recovery attempts.
    /// </summary>
    public int RecoveryRetries { get; }

    /// <summary>
    /// Server lack-of-ack timeout in milliseconds.
    /// </summary>
    public int ServerLackOfAckTimeoutMs { get; }

    /// <summary>
    /// Flush timeout in milliseconds.
    /// </summary>
    public int FlushTimeoutMs { get; }

    /// <summary>
    /// Connection timeout in milliseconds.
    /// </summary>
    public int ConnectionTimeoutMs { get; }

    /// <summary>
    /// IPC compression type. Default: None.
    /// </summary>
    public IPCCompressionType IpcCompression { get; }

    /// <summary>
    /// Maximum wait time when the stream is paused, in milliseconds.
    /// A negative value means "wait the full server-specified duration."
    /// </summary>
    public long StreamPausedMaxWaitTimeMs { get; }

    private ArrowStreamConfigurationOptions(Builder builder)
    {
        MaxInflightBatches = builder.MaxInflightBatches;
        Recovery = builder.Recovery;
        RecoveryTimeoutMs = builder.RecoveryTimeoutMs;
        RecoveryBackoffMs = builder.RecoveryBackoffMs;
        RecoveryRetries = builder.RecoveryRetries;
        ServerLackOfAckTimeoutMs = builder.ServerLackOfAckTimeoutMs;
        FlushTimeoutMs = builder.FlushTimeoutMs;
        ConnectionTimeoutMs = builder.ConnectionTimeoutMs;
        IpcCompression = builder.IpcCompression;
        StreamPausedMaxWaitTimeMs = builder.StreamPausedMaxWaitTimeMs;
    }

    /// <summary>
    /// Returns the default Arrow stream configuration options.
    /// </summary>
    public static ArrowStreamConfigurationOptions Default => new(new Builder());

    /// <summary>
    /// Creates a new builder for ArrowStreamConfigurationOptions.
    /// </summary>
    public static Builder NewBuilder() => new();

    /// <summary>
    /// Builder for creating <see cref="ArrowStreamConfigurationOptions"/> instances.
    /// </summary>
    public sealed class Builder
    {
        internal int MaxInflightBatches { get; set; } = DefaultMaxInflightBatches;
        internal bool Recovery { get; set; } = StreamConfigurationOptions.DefaultRecovery;
        internal int RecoveryTimeoutMs { get; set; } = StreamConfigurationOptions.DefaultRecoveryTimeoutMs;
        internal int RecoveryBackoffMs { get; set; } = StreamConfigurationOptions.DefaultRecoveryBackoffMs;
        internal int RecoveryRetries { get; set; } = StreamConfigurationOptions.DefaultRecoveryRetries;
        internal int ServerLackOfAckTimeoutMs { get; set; } = StreamConfigurationOptions.DefaultServerLackOfAckTimeoutMs;
        internal int FlushTimeoutMs { get; set; } = StreamConfigurationOptions.DefaultFlushTimeoutMs;
        internal int ConnectionTimeoutMs { get; set; } = DefaultConnectionTimeoutMs;
        internal IPCCompressionType IpcCompression { get; set; } = IPCCompressionType.None;
        internal long StreamPausedMaxWaitTimeMs { get; set; } = DefaultStreamPausedMaxWaitTimeMs;

        internal Builder() { }

        /// <summary>
        /// Sets the maximum number of in-flight Arrow batches. Must be positive.
        /// </summary>
        public Builder SetMaxInflightBatches(int maxInflightBatches)
        {
            if (maxInflightBatches <= 0)
                throw new ArgumentException("maxInflightBatches must be positive", nameof(maxInflightBatches));
            MaxInflightBatches = maxInflightBatches;
            return this;
        }

        /// <summary>
        /// Enables or disables automatic recovery.
        /// </summary>
        public Builder SetRecovery(bool recovery) { Recovery = recovery; return this; }

        /// <summary>
        /// Sets the recovery timeout in milliseconds. Must be non-negative.
        /// </summary>
        public Builder SetRecoveryTimeoutMs(int ms)
        {
            if (ms < 0) throw new ArgumentException("Must be non-negative", nameof(ms));
            RecoveryTimeoutMs = ms; return this;
        }

        /// <summary>
        /// Sets the recovery backoff in milliseconds. Must be non-negative.
        /// </summary>
        public Builder SetRecoveryBackoffMs(int ms)
        {
            if (ms < 0) throw new ArgumentException("Must be non-negative", nameof(ms));
            RecoveryBackoffMs = ms; return this;
        }

        /// <summary>
        /// Sets the maximum number of recovery retries. Must be non-negative.
        /// </summary>
        public Builder SetRecoveryRetries(int retries)
        {
            if (retries < 0) throw new ArgumentException("Must be non-negative", nameof(retries));
            RecoveryRetries = retries; return this;
        }

        /// <summary>
        /// Sets the server lack-of-ack timeout in milliseconds. Must be positive.
        /// </summary>
        public Builder SetServerLackOfAckTimeoutMs(int ms)
        {
            if (ms <= 0) throw new ArgumentException("Must be positive", nameof(ms));
            ServerLackOfAckTimeoutMs = ms; return this;
        }

        /// <summary>
        /// Sets the flush timeout in milliseconds. Must be positive.
        /// </summary>
        public Builder SetFlushTimeoutMs(int ms)
        {
            if (ms <= 0) throw new ArgumentException("Must be positive", nameof(ms));
            FlushTimeoutMs = ms; return this;
        }

        /// <summary>
        /// Sets the connection timeout in milliseconds. Must be positive.
        /// </summary>
        public Builder SetConnectionTimeoutMs(int ms)
        {
            if (ms <= 0) throw new ArgumentException("Must be positive", nameof(ms));
            ConnectionTimeoutMs = ms; return this;
        }

        /// <summary>
        /// Sets the IPC compression type.
        /// </summary>
        public Builder SetIpcCompression(IPCCompressionType compression)
        {
            IpcCompression = compression; return this;
        }

        /// <summary>
        /// Sets the maximum wait time when the stream is paused.
        /// A negative value means "wait the full server-specified duration."
        /// </summary>
        public Builder SetStreamPausedMaxWaitTimeMs(long ms)
        {
            StreamPausedMaxWaitTimeMs = ms; return this;
        }

        /// <summary>
        /// Builds the <see cref="ArrowStreamConfigurationOptions"/> instance.
        /// </summary>
        public ArrowStreamConfigurationOptions Build() => new(this);
    }
}
